### PR TITLE
Avoid reimporting lightmap textures every getter call

### LINKED ISSUE #77777: LightmapGI : Duplicating any LightmapGIData crashes the engine
### Godot version

v4.1.dev.custom_build [2e273f0e35800f2bcc7c05db54100caa97928f69]

### System information

Windows 11 x64, Vulkan (Forward+), NVIDIA GeForce GTX 1650

### Issue description

Duplicating `LightmapGIData` crashes the engine. Only caused when running the project. Not in the editor.

Traceback :
```
 	0000000000000000()	Unknown
>	godot.windows.editor.dev.x86_64.exe!LightmapGIData::_get_light_textures_data() Line 179	C++
 	godot.windows.editor.dev.x86_64.exe!call_with_variant_args_retc_helper<LightmapGIData,Array>(LightmapGIData * p_instance, Array(const LightmapGIData::*)() p_method, const Variant * * p_args, Variant & r_ret, Callable::CallError & r_error, IndexSequence<> __formal) Line 806	C++
 	godot.windows.editor.dev.x86_64.exe!call_with_variant_args_retc_dv<LightmapGIData,Array>(LightmapGIData * p_instance, Array(const LightmapGIData::*)() p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error, const Vector<Variant> & default_values) Line 568	C++
 	godot.windows.editor.dev.x86_64.exe!MethodBindTRC<LightmapGIData,Array>::call(Object * p_object, const Variant * * p_args, int p_arg_count, Callable::CallError & r_error) Line 581	C++
 	godot.windows.editor.dev.x86_64.exe!ClassDB::get_property(Object * p_object, const StringName & p_property, Variant & r_value) Line 1184	C++
 	godot.windows.editor.dev.x86_64.exe!Object::get(const StringName & p_name, bool * r_valid) Line 343	C++
 	godot.windows.editor.dev.x86_64.exe!Resource::duplicate(bool p_subresources) Line 261	C++
 	godot.windows.editor.dev.x86_64.exe!call_with_variant_args_retc_helper<Resource,Ref<Resource>,bool,0>(Resource * p_instance, Ref<Resource>(const Resource::*)(bool) p_method, const Variant * * p_args, Variant & r_ret, Callable::CallError & r_error, IndexSequence<0> __formal) Line 806	C++
 	godot.windows.editor.dev.x86_64.exe!call_with_variant_args_retc_dv<Resource,Ref<Resource>,bool>(Resource * p_instance, Ref<Resource>(const Resource::*)(bool) p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error, const Vector<Variant> & default_values) Line 568	C++
 	godot.windows.editor.dev.x86_64.exe!MethodBindTRC<Resource,Ref<Resource>,bool>::call(Object * p_object, const Variant * * p_args, int p_arg_count, Callable::CallError & r_error) Line 581	C++
 	godot.windows.editor.dev.x86_64.exe!Object::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 739	C++
 	godot.windows.editor.dev.x86_64.exe!Variant::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Variant & r_ret, Callable::CallError & r_error) Line 1174	C++
 	godot.windows.editor.dev.x86_64.exe!GDScriptFunction::call(GDScriptInstance * p_instance, const Variant * * p_args, int p_argcount, Callable::CallError & r_err, GDScriptFunction::CallState * p_state) Line 1662	C++
 	godot.windows.editor.dev.x86_64.exe!GDScriptInstance::callp(const StringName & p_method, const Variant * * p_args, int p_argcount, Callable::CallError & r_error) Line 1787	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_gdvirtual__ready_call<0>() Line 322	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_notification(int p_notification) Line 192	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_notificationv(int p_notification, bool p_reversed) Line 49	C++
 	godot.windows.editor.dev.x86_64.exe!Node3D::_notificationv(int p_notification, bool p_reversed) Line 52	C++
 	godot.windows.editor.dev.x86_64.exe!Object::notification(int p_notification, bool p_reversed) Line 798	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_propagate_ready() Line 244	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_propagate_ready() Line 235	C++
 	godot.windows.editor.dev.x86_64.exe!Node::_set_tree(SceneTree * p_tree) Line 2913	C++
 	godot.windows.editor.dev.x86_64.exe!SceneTree::initialize() Line 449	C++
 	godot.windows.editor.dev.x86_64.exe!OS_Windows::run() Line 1476	C++
 	godot.windows.editor.dev.x86_64.exe!widechar_main(int argc, wchar_t * * argv) Line 181	C++
 	godot.windows.editor.dev.x86_64.exe!_main() Line 203	C++
 	godot.windows.editor.dev.x86_64.exe!main(int argc, char * * argv) Line 217	C++
 	godot.windows.editor.dev.x86_64.exe!WinMain(HINSTANCE__ * hInstance, HINSTANCE__ * hPrevInstance, char * lpCmdLine, int nCmdShow) Line 231	C++
```

### Steps to reproduce

1. Baking lightmaps in any scenes
2. Trying to duplicate a LightmapGIData resource in run-time
Example using GDScript :
```GDScript
func _ready() :
    $lightmapgi.light_data.duplicate()
```
3. Run the project
4. Crash

### Minimal reproduction project

Doesn't need MRP. Just do the steps above

### LINKED ISSUE #83444: Moving nodes around the tree in a scene with baked lightmap triggers lightmap reimport
### Godot version

v4.2.beta.custom_build [a60fc7f7c]

### System information

Godot v4.2.beta (a60fc7f7c) - Windows 10.0.22621 - Vulkan (Forward+) - dedicated AMD Radeon RX 6700 XT (Advanced Micro Devices, Inc.; 27.20.21034.37) - AMD Ryzen 7 5700X 8-Core Processor (16 Threads)

### Issue description

When you move any node around within the tree of a scene with baked lightmap, a reimport of the lightmap file will be triggered.

### Steps to reproduce

1. Set up a scene with some lightmapped meshes. Also create one dynamically lit object for testing reason.
2. Bake lightmap
3. Move any node around in the scene tree
4. A reimport will be triggered.

### Minimal reproduction project

[Lightmap Reimport.zip](https://github.com/godotengine/godot/files/12918646/Lightmap.Reimport.zip)

### PR DESCRIPTION
**Pull request by gongpha**
- Fixes #77777. (I didn't notice that was a fantastic number - -)
- *Bugsquad edit:* Fixes https://github.com/godotengine/godot/issues/83444

Also improves #61861. The `_get_light_textures_data` function is trying to reimport lightmap textures on every getter call which sounds irritable to me (e.g. when the engine is reading the property via duplicating).

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally (rebased on top of `master` 329652b6feac2114c4e30062f18ece36d4e3addc), it works as expected. I no longer get a crash with the MRP.

However, we will need to implement a compatibility handler so that users aren't forced to rebake lightmaps when upgrading. (This can take a long time on complex levels with high quality settings, especially if you have a slow GPU.)

**Comment by akien-mga** [comment]
This indeed needs some compatibility code added. The old methods should still be registered as compatibility methods for GDExtension, and existing data should ideally be parsed and converted to the new format (which can be done via the `_set`/`_get` methods).

File: scene/3d/lightmap_gi.cpp [hunk 1]
```diff
@@ -1085,6 +1021,69 @@ LightmapGI::BakeError LightmapGI::bake(Node *p_from_node, String p_image_data_pa
 		return BAKE_ERROR_MESHES_INVALID;
 	}
 
+	/* POSTBAKE: Save Textures */
```
[the remark this answers is not in the document]
Comment (gongpha) [reply scene/3d/lightmap_gi.cpp]: Forgot to note, These comments (including the ones below too) were written from the code before #61861 would remove them. I just re-added them back. And there are also more unformatted comments in this file. Should I edit them too ?

[the remark this answers is not in the document]
Comment (AThousandShips) [reply scene/3d/lightmap_gi.cpp]: Don't touch the rest, it applies for comments in your PR, including copied or moved, it's to keep the edit history clean, not changing areas that aren't affected, but also making it slowly up to date

**Comment by atirut-w** [comment]
I don't know what went wrong during the time I last tested this branch, but it no longer works. It crashes with an out of bound error seemingly when there are more than one lightmapped meshes in a scene.

```
...
ERROR: FATAL: Index p_index = 1 is out of bounds (size() = 1).
   at: CowData<class Ref<class Image> >::get (C:\Users\atiru\Projects\godot\core/templates/cowdata.h:155)

================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.dev.custom_build (865c25adfc9f9842dde9aaf3162f1b768de57c6c)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] CowData<Ref<Image> >::get (C:\Users\atiru\Projects\godot\core\templates\cowdata.h:155)
[1] Vector<Ref<Image> >::operator[] (C:\Users\atiru\Projects\godot\core\templates\vector.h:96)
[2] LightmapGI::bake (C:\Users\atiru\Projects\godot\scene\3d\lightmap_gi.cpp:1088)
[3] LightmapGIEditorPlugin::_bake_select_file (C:\Users\atiru\Projects\godot\editor\plugins\lightmap_gi_editor_plugin.c)
[4] LightmapGIEditorPlugin::_bake (C:\Users\atiru\Projects\godot\editor\plugins\lightmap_gi_editor_plugin.cpp:113)
...
```

MRP (don't mind the namings, I reused an old MRP): [crash.zip](https://github.com/godotengine/godot/files/13062354/crash.zip)

**Comment by gongpha** [comment]
This crash issue seems not related to this PR. I only refactored the old porting PR (#61861) and removed unnecessary importing operations. You may create a new issue and fill in details about the issue and MRP.

**Review by AThousandShips — APPROVED** [review]

**Comment by atirut-w** [comment]
This issue does not happen in c5a7462a00a665a72c7c1a746c105e3c1c475c96, which is the parent of this PR's branch. That should confirm that this issue is caused by something during all those rebases.

File: scene/3d/lightmap_gi.cpp [hunk 2]
```diff
@@ -1086,6 +1056,69 @@ LightmapGI::BakeError LightmapGI::bake(Node *p_from_node, String p_image_data_pa
...
+
+		int slice_count = images.size();
+		int slice_width = images[0]->get_width();
+		int slice_height = images[0]->get_height();
+
+		int slices_per_texture = Image::MAX_HEIGHT / slice_height;
+		int texture_count = Math::ceil(slice_count / (float)slices_per_texture);
+
+		textures.resize(texture_count);
+		images.resize(texture_count);
```
Comment (AThousandShips) [inline scene/3d/lightmap_gi.cpp:1077]: I think this is the cause of the error, this shouldn't be resized, right?

**Review by AThousandShips — DISMISSED** [review]
See above

File: scene/3d/lightmap_gi.cpp (hunk 2 above)
Comment (gongpha) [reply scene/3d/lightmap_gi.cpp:1077]: Probably. I don't remember how this line was added :/

**Comment by atirut-w** [comment]
@AThousandShips I tried out your fix on my end and it still crashes with the exact same log, but seems to have progressed a little more. Before, it only said "Retrieving textures", but it now shows reimport progress of the lightmap.

A finally decided to actually test it with a debugger. Here's what I got:

On line 1088: `texture_image->blit_rect(images[i * slices_per_texture + j], Rect2i(0, 0, slice_width, slice_height), Point2i(0, slice_height * j));`

`images` is an array with one member in my case BUT is multiplied with other things which resulted in a bogus index of `1`.

```
> images[0]
{reference=0x000001baca91e2b0 {format=FORMAT_RGBH (14) data={[size]=6144 [0]=167 'ง' [1]=48 '0' ...} ...} }
> images[1]
Array index out of bounds.
```

```
> i * slices_per_texture + j
1
```

**Review by AThousandShips — DISMISSED** [review]
Needs further evaluation, ensure that the code is correctly transferred and correct

**Comment by atirut-w** [comment]
New version works now. Turns out I applied it wrong by also commenting out `textures.resize(texture_count);`.

**Comment by atirut-w** [comment]
Possibly offtopic, but I noticed this issue was on 4.3 milestone. That was such a mistake.

**Comment by AThousandShips** [comment]
Indeed off topic

**Review by clayjohn — APPROVED** [review]
Makes sense to me. Kobewi confirmed in chat that it looks fine as well. I'm comfortable going ahead with this

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**